src/python/pants/option/config.py

```diff
@@ -198,40 +260,30 @@ def get_source_for_option(self, section: str, option: str) -> None:
     return None
 
 
+@dataclass(frozen=True)
 class _SingleFileConfig(Config):
-  """Config read from a single file.
-
-  NB: In order to have:
-    1. a specialized implementation of __eq__ and __hash__ that avoids comparing file contents
-    2. equality ignore the ConfigParser instance
-  ...this is not a dataclass.
```

Comment:
Turns out, that wasn't really relevant. We can still override `__eq__` and `__hash__` when using a dataclass. Making the change gives us some benefits:
1) Less boilerplate for `__init__`
2) Instance is now frozen
3) Better `__str__`.